File: LeetCode 75/Two Pointers/11. Container With Most Water.test.js
/*

ou are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

Find two lines that together with the x-axis form a container, such that the container contains the most water.

Return the maximum amount of water a container can store.

Notice that you may not slant the container.



Example 1:


Input: height = [1,8,6,2,5,4,8,3,7]
Output: 49
Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.
Example 2:

Input: height = [1,1]
Output: 1


Constraints:

n == height.length
2 <= n <= 105
0 <= height[i] <= 104

*/

/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
    let num1 = Number.NEGATIVE_INFINITY,
        num2 = Number.NEGATIVE_INFINITY,
        num1Index = -1,
        num2Index = -1,
        areaSize = 0;

    for (let i = 0; i < height.length; i++) {
        if (height[i] > num1) {
            num1 = height[i];
            num1Index = i;
        } else if (height[i] > num2 || (height[i] * (i - num1Index) > areaSize)) {
            num2 = height[i];
            num2Index = i;
            areaSize = num2 * (num2Index - num1Index)
        }
    }

    if (num1 > 0 && num2 > 0) {
        return areaSize;
    }

    return 0;
};

describe('11. Container With Most Water', () => {
    test('[1, 8, 6, 2, 5, 4, 8, 3, 7]', () => expect(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7])).toEqual(49));
    test('[1, 1]', () => expect(maxArea([1, 1])).toEqual(1));
    test('[]', () => expect(maxArea([])).toEqual(0));
    test('[1, 2, 1]', () => expect(maxArea([1, 2, 1])).toEqual(2));
})

File: solutions/0000-0099/11. Container With Most Water_test.go
package main

import "testing"

/*

ou are given an integer array height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).

Find two lines that together with the x-axis form a container, such that the container contains the most water.

Return the maximum amount of water a container can store.

Notice that you may not slant the container.



Example 1:


Input: height = [1,8,6,2,5,4,8,3,7]
Output: 49
Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.
Example 2:

Input: height = [1,1]
Output: 1


Constraints:

n == height.length
2 <= n <= 10^5
0 <= height[i] <= 10^4

*/

// maxArea finds the maximum area of water that can be stored between two lines.
// Uses the two-pointer technique to efficiently find the optimal solution.
//
// The algorithm:
// 1. Start with pointers at both ends of the array
// 2. Calculate area formed by the two pointers
// 3. Move the pointer pointing to the shorter line inward
// 4. Continue until pointers meet
//
// This works because moving the taller line's pointer can only decrease the area,
// while moving the shorter line's pointer might find a taller line that results
// in a larger area despite the decreased width.
//
// Parameters:
//
//	height []int - Array representing heights of vertical lines
//
// Returns:
//
//	int - Maximum area of water that can be stored
func maxArea(height []int) int {
	// Handle edge case: need at least 2 lines to form a container
	if len(height) < 2 {
		return 0
	}

	areaSize := 0
	left := 0
	right := len(height) - 1

	// Two-pointer approach
	for left < right {
		current := min(height[left], height[right]) * (right - left)
		areaSize = max(areaSize, current)

		// Move the pointer pointing to the shorter line
		// This is the key insight: moving the taller line can only decrease area
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return areaSize
}

func TestMaxArea(t *testing.T) {
	const name = "11. Container With Most Water"
	// Array of 10000 ones
	ones := make([]int, 10000)
	for i := range ones {
		ones[i] = 1
	}
	testCases := []struct {
		height []int
		want   int
	}{
		{height: []int{1, 8, 6, 2, 5, 4, 8, 3, 7}, want: 49},
		{height: []int{1, 1}, want: 1},
		{height: []int{0}, want: 0},
		{height: []int{1, 2, 1}, want: 2},
		{height: []int{1, 2, 4, 3}, want: 4},
		{height: []int{2, 3, 4, 5, 18, 17, 6}, want: 17},
		{height: []int{0, 14, 6, 2, 10, 9, 4, 1, 10, 3}, want: 70},
		{height: ones, want: 9999},
	}
	for _, tc := range testCases {
		t.Run(name, func(t *testing.T) {
			out := maxArea(tc.height)
			if out != tc.want {
				t.Errorf("maxArea(%v) = %v, want: %v", tc.height, out, tc.want)
			}
		})
	}
}
